### DISCUSSION TITLE
xmake-js or xmake-python?

### DISCUSSION THREAD

**Discussion by Freed-Wu** [xmake-io/xmake#5125] [Ideas]
We have [cmake-js](https://github.com/cmake-js/cmake-js) to create nodejs module by cmake.
We also have [meson-python](https://github.com/mesonbuild/meson-python) to create python module by meson.

Perhaps we can have xmake-js and xmake-python?

**Comment by star-hengxing** [comment]
If anyone is interested, open a PR and do it

**Comment by Fuatorium** [comment]
Yes, creating xmake-js and xmake-python would facilitate building Node.js and Python modules using Xmake, adding value to the existing ecosystem.

**Reply by Freed-Wu** [reply]
Now,

- anyone who want to use xmake to create nodejs package can see https://github.com/xmake-io/xmake-wheel/blob/main/examples/nodejs/1_hello_world
- anyone who want to use xmake to create python package can see https://github.com/xmake-io/xmake-python/tree/1c260155608a0d6208c27ad7183714ee4eb76821/tests/examples/xmake/our-first-project

**Comment by Freed-Wu** [comment]
https://github.com/solarwinds/zig-build is an npm package to use zig to build nodejs addon.
It looks like xmake-js should do two things: Download xmake according platform, and refer https://stackoverflow.com/questions/58633310/how-to-compile-node-js-native-api-extension-via-meson-build-system
to build a nodejs addon on `build/release/XXX.node`.

BTW, npm builtin build system node-gyp has bugs on android and macos, and don't support mingw. I think a well-designed xmake can solve these problems.

**Comment by Freed-Wu** [comment]
xmake support for nodejs module will be fixed by #5949 and #5691 (npm).
xmake support for python module still doesn't have plan. #5691(PYPI), cython compiler, cpython module
and for lua module, #5691 (luarocks) is helpful.

**Reply by Freed-Wu** [reply]
BTW, what is your PYPI username @waruqi ? https://pypi.org/project/xmake-python/

**Reply by waruqi** [reply]
waruqi
